src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -795,78 +796,152 @@ private string GetPreferredEnvironmentHost(string environmentHost, InstanceDisco
-                    adalUsersResult = CacheFallbackOperations.GetAllAdalUsersForMsal(_logger, LegacyCachePersistence, ClientId);
+
+                    // Known Issue: will ignore env aliases, which we cannot get without an instance discovery
+                    adalUsersResult = CacheFallbackOperations.GetAllAdalUsersForMsal(
+                        _logger,
+                        LegacyCachePersistence,
+                        ClientId,
+                        environment);
                 }
                 finally
                 {
                     OnAfterAccess(args);
                 }
+            }
+        }
 
-                IDictionary<string, Account> clientInfoToAccountMap = new Dictionary<string, Account>();
-                foreach (MsalRefreshTokenCacheItem rtItem in tokenCacheItems)
-                {
-                    foreach (MsalAccountCacheItem account in accountCacheItems)
-                    {
-                        if (rtItem.HomeAccountId.Equals(account.HomeAccountId, StringComparison.OrdinalIgnoreCase))
-                        {
-                            clientInfoToAccountMap[rtItem.HomeAccountId] = new Account
-                                (account.HomeAccountId, account.PreferredUsername, environment);
-                            break;
-                        }
-                    }
-                }
+        private async Task<string> GetPreferedCacheEnvAvoidNetworkCallAsync(
+            string authority,
+            IEnumerable<MsalAccountCacheItem> accountCacheItems)
+        {
+            // this map is to try to avoid a network call when getting accounts and it is ok if it is not up to date
+            // but it should not be used in any other scenario
+            IDictionary<string, string> envToPreferredCache = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
+            {
+                { AzurePublicEnv, "login.windows.net" },
+                { "login.partner.microsoftonline.cn", "login.partner.microsoftonline.cn" },
+                { "login.microsoftonline.de", "login.microsoftonline.de" },
+                { "login.microsoftonline.us", "login.microsoftonline.us" },
+                { "login-us.microsoftonline.com", "login-us.microsoftonline.com" },
+            };
 
-                Dictionary<string, AdalUserInfo> clientInfoToAdalUserMap = adalUsersResult.ClientInfoUsers;
-                List<AdalUserInfo> adalUsersWithoutClientInfo = adalUsersResult.UsersWithoutClientInfo;
+            var envFromRequest = Authority.GetEnviroment(authority);
 
-                foreach (KeyValuePair<string, AdalUserInfo> pair in clientInfoToAdalUserMap)
-                {
-                    ClientInfo clientInfo = ClientInfo.CreateFromJson(pair.Key);
-                    string accountIdentifier = clientInfo.ToAccountIdentifier();
+            // Avoid a network call if all accounts are from the same network (no aliases!)
```

Comment:
>network [](start = 70, length = 7)

what do you mean "same network"? like same environment?

Reply:
Yes, same enviroment. Fixed. Thx for pointing this out.